clientapi/routing/consent_tracking.go

```diff
@@ -0,0 +1,226 @@
+// Copyright 2022 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package routing
+
+import (
+	"bytes"
+	"context"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"net/http"
+
+	appserviceAPI "github.com/matrix-org/dendrite/appservice/api"
+	"github.com/matrix-org/dendrite/clientapi/jsonerror"
+	"github.com/matrix-org/dendrite/roomserver/api"
+	"github.com/matrix-org/dendrite/setup/config"
+	userapi "github.com/matrix-org/dendrite/userapi/api"
+	userdb "github.com/matrix-org/dendrite/userapi/storage"
+	"github.com/matrix-org/gomatrixserverlib"
+	"github.com/matrix-org/util"
+	"github.com/sirupsen/logrus"
+)
+
+// The data used to populate the /consent request
+type constentTemplateData struct {
+	User          string
+	Version       string
+	UserHMAC      string
+	HasConsented  bool
+	PublicVersion bool
+}
+
+func consent(writer http.ResponseWriter, req *http.Request, userAPI userapi.UserInternalAPI, cfg *config.ClientAPI) *util.JSONResponse {
+	consentCfg := cfg.Matrix.UserConsentOptions
+	internalError := jsonerror.InternalServerError()
+
+	// The data used to populate the /consent request
+	data := constentTemplateData{
+		User:     req.FormValue("u"),
+		Version:  req.FormValue("v"),
+		UserHMAC: req.FormValue("h"),
+	}
+	switch req.Method {
+	case http.MethodGet:
+		// display the privacy policy without a form
+		data.PublicVersion = data.User == "" || data.UserHMAC == "" || data.Version == ""
```

Comment:
`PublicVersion` isn't a great name. AFAICT it seems to control whether there is an accept button or not. So perhaps "read only" would be better.